### PR TITLE
feat(mstsgu): add smart-card gateway authentication

### PR DESCRIPTION
**Pull request by mamoreau-devolutions**
Add an opt-in Kerberos PKINIT path for HTTP Negotiate gateway authentication while retaining the password Negotiate, NTLM, and Basic flows.

The public credentials type accepts an application-supplied UPN, redacts credentials, and rejects unsupported smart-card feature or challenge combinations without exposing them.

### CODE REVIEW

File: crates/ironrdp-mstsgu/tests/http_auth.rs [hunk 1]
```diff
@@ -58,32 +72,83 @@ fn ntlm_type1_from_challenge() {
...
+    let Err(error) = result else {
+        panic!("smart-card authentication requires Negotiate");
+    };
+    let error = error.to_string();
+
+    assert!(error.contains("Negotiate"));
+    assert!(!error.contains("sensitive-pin"));
+    assert!(!error.contains("sensitive-certificate"));
+    assert!(!error.contains("sensitive-private-key"));
+}
```
[the remark this answers is not in the document]
Comment (mamoreau-devolutions) [reply crates/ironrdp-mstsgu/tests/http_auth.rs:169]: Fixed in 5d65d597. `smart_card_recognizes_combined_negotiate_challenge` supplies a combined `NTLM`, `Negotiate`, and `Basic` header and verifies that malformed Negotiate input is rejected rather than downgraded.

> [!NOTE]
> Human-tuned, LLM-assisted content.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: crates/ironrdp-mstsgu/tests/http_auth.rs [hunk 2]
```diff
@@ -58,32 +72,83 @@ fn ntlm_type1_from_challenge() {
...
+    assert!(error.contains("smart-card"));
+    assert!(error.contains("smartcard"));
+    assert!(!error.contains("sensitive-pin"));
+    assert!(!error.contains("sensitive-certificate"));
+    assert!(!error.contains("sensitive-private-key"));
+}
+
+#[cfg(feature = "smartcard")]
+#[test]
+fn smart_card_requires_negotiate() {
```
[the remark this answers is not in the document]
Comment (mamoreau-devolutions) [reply crates/ironrdp-mstsgu/tests/http_auth.rs:157]: Fixed CI coverage in 5d65d597 by adding `native-tls,smartcard` compile and execution invocations. The focused tests also cover malformed first-round and combined `Negotiate` challenges without starting PKINIT or a network operation. I intentionally did not add a certificate/key fixture because this change's stated test boundary prohibits certificate, credential, and key material in tests.

> [!NOTE]
> Human-tuned, LLM-assisted content.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: crates/ironrdp-mstsgu/src/http_auth.rs [hunk 3]
```diff
@@ -191,6 +217,108 @@ impl GatewayHttpAuth {
...
+            })
+            .and_then(|asn1| picky_asn1_der::from_bytes(&asn1.0.0).ok())
+            .or_else(|| {
+                certificate
+                    .tbs_certificate
+                    .subject
+                    .find_common_name()
+                    .map(ToString::to_string)
+            })
+    }
```
[the remark this answers is not in the document]
Comment (mamoreau-devolutions) [reply crates/ironrdp-mstsgu/src/http_auth.rs]: I did not add a certificate or private-key test fixture because the task explicitly prohibits certificate, credential, and key material in tests. The implementation keeps the existing SSPI/Picky parsing path feature-gated; the focused coverage validates the non-secret authentication-selection and challenge behavior without opening a KDC connection.

> [!NOTE]
> Human-tuned, LLM-assisted content.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: crates/ironrdp-mstsgu/src/lib.rs [hunk 4]
```diff
@@ -389,7 +430,7 @@ async fn websocket_upgrade_with_auth(
         } else {
             let (auth, step) = run_http_auth(move || {
                 let refs: Vec<&str> = challenges.iter().map(String::as_str).collect();
-                GatewayHttpAuth::from_challenges(&user, &pass, Some(target_name), &refs)
+                GatewayHttpAuth::from_challenges(&user, &pass, smart_card.as_deref(), Some(target_name), &refs)
```
[the remark this answers is not in the document]
Comment (mamoreau-devolutions) [reply crates/ironrdp-mstsgu/src/lib.rs]: Fixed in 5d65d597. Smart-card Negotiate state disables Basic fallback, and the WebSocket upgrade loop separately rejects a `TryBasic` result while a smart card is selected. The password flow keeps its existing Basic fallback.

> [!NOTE]
> Human-tuned, LLM-assisted content.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: crates/ironrdp-mstsgu/src/http_auth.rs [hunk 5]
```diff
@@ -60,9 +60,35 @@ impl GatewayHttpAuth {
...
+        #[cfg(feature = "smartcard")]
+        if let Some(smart_card) = smart_card {
+            if iter_auth_challenges(challenges.iter().copied())
+                .any(|challenge| split_auth_challenge(challenge, "Negotiate").is_some())
+            {
+                let mut auth = Self::new_negotiate_smartcard(smart_card, target_name)?;
+                let token = auth.initialize(None)?;
+                let header = auth.format_authorization(&token);
+                return Ok((Some(auth), AuthStep::Continue(header)));
+            }
```
[the remark this answers is not in the document]
Comment (mamoreau-devolutions) [reply crates/ironrdp-mstsgu/src/http_auth.rs]: Fixed in 5d65d597. Smart-card authentication now locates, base64-decodes, and passes the first `Negotiate` challenge token to `initialize`; malformed tokens return the decode error before PKINIT initialization.

> [!NOTE]
> Human-tuned, LLM-assisted content.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: crates/ironrdp-mstsgu/src/packet_io.rs [hunk 6]
```diff
@@ -189,6 +191,12 @@ async fn websocket_upgrade_with_auth(
 
         match step {
             AuthStep::Continue(next) => authorization = Some(next),
+            AuthStep::TryBasic if using_smart_card => {
+                return Err(Error::new(
+                    "gateway does not continue Negotiate smart-card authentication",
+                    GwErrorKind::UnsupportedFeature,
+                ));
+            }
```
[the remark this answers is not in the document]
Comment (mamoreau-devolutions) [reply crates/ironrdp-mstsgu/src/packet_io.rs]: Removed in 62033bc0. Smart-card auth permanently disables Basic fallback in `GatewayHttpAuth`, so the downstream guard was unreachable.

> [!NOTE]
> Human-tuned, LLM-assisted content.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Comment by mamoreau-devolutions** [comment]
Addressed in 62033bc0. `GwSmartCardCredentials` now requires an application-supplied Kerberos UPN, which removes the duplicate certificate SAN/CN identity inference entirely. Reusing the connector helper would make this HTTP-only feature depend on the CredSSP connector stack, contrary to the requested scope; certificate parsing remains only for the PKINIT material supplied to SSPI.

> [!NOTE]
> Human-tuned, LLM-assisted content.
